File: Auto_Use/mac/llm_provider/together/service.py
# Copyright 2026 Cursortouch — Auto-Use

import json
import requests
from typing import Dict, Any, Optional
from .. import LLM_HTTP_TIMEOUT

class TogetherProvider:
    """Together AI API provider for LLM interactions (OpenAI-compatible chat completions)"""

    def __init__(self, api_key: str, cli_agent: bool = False, model_info: dict = None, tools: list = None):
        self.api_key = api_key
        # api.together.ai is an alias of the same endpoint.
        self.api_url = "https://api.together.xyz/v1/chat/completions"
        self.cli_agent = cli_agent
        self.model_info = model_info or {}
        # Native tool calling: OpenAI-format function tools — the model's
        # output contract. None only for mode="text" (plain prose).
        self.tools = tools or None
        # Together documents tool_choice as auto|none|{function}; "required"
        # is undocumented. Start with "required" (the loop's contract) and
        # downgrade to "auto" for the rest of the session only once Together
        # proves it rejects "required" — see send_request.
        self._tool_choice = "required"

    def send_request(self, messages: list, model: str, annotated_screenshot_base64: Optional[str] = None) -> Dict[str, Any]:
        """Send request to Together AI API"""

        # If screenshot is provided and NOT cli_agent, modify the user message to include the annotated image
        if annotated_screenshot_base64 and not self.cli_agent and len(messages) > 1:
            user_message = messages[-1]["content"]
            messages[-1]["content"] = [
                {
                    "type": "text",
                    "text": user_message
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": f"data:image/jpeg;base64,{annotated_screenshot_base64}"
                    }
                }
            ]

        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }

        data = {
            "model": model,
            "messages": messages,
            "temperature": 0.2,
            # Reasoning models bill their thinking as completion tokens; 4000
            # risks finish_reason "length" before the tool call is emitted.
            "max_tokens": 10000,
        }

        # No tools (mode="text") -> plain text: omit the tool params entirely.
        if self.tools:
            data["tools"] = self.tools
            # "required" - every turn must call at least one tool; a text-only
            # turn is never a valid step (even termination is the `exit`
            # tool). Canonical rationale: see openrouter/service.py.
            data["tool_choice"] = self._tool_choice

        try:
            response = requests.post(self.api_url, json=data, headers=headers, timeout=LLM_HTTP_TIMEOUT)
            if response.status_code == 400 and self.tools and self._tool_choice == "required":
                # Retry once with "auto". Remember it ONLY if that works — an
                # unrelated 400 falls through and raises with its own body.
                # The agent loop repairs the odd text-only turn ("no tool
                # called"), so "auto" degrades gracefully.
                data["tool_choice"] = "auto"
                retry = requests.post(self.api_url, json=data, headers=headers, timeout=LLM_HTTP_TIMEOUT)
                if retry.ok:
                    self._tool_choice = "auto"
                    response = retry
            response.raise_for_status()
            result = response.json()
            if self.tools:
                return _normalize_tool_response(result)
            return result
        except requests.exceptions.RequestException as e:
            error_msg = f"Together API request failed: {str(e)}"
            if hasattr(e, 'response') and e.response is not None:
                error_msg += f"\nResponse: {e.response.text}"
            raise Exception(error_msg)


def _normalize_tool_response(result: dict) -> dict:
    """Normalize an OpenAI-format tool-call response: keep the usual
    choices/message shape but replace tool_calls with
    [{"id", "name", "arguments": dict}] (arguments JSON-decoded). The `id` is
    echoed back by the loop so each tool result matches its call."""
    message = ((result.get("choices") or [{}])[0].get("message")) or {}
    calls = []
    for i, tc in enumerate(message.get("tool_calls") or []):
        fn = (tc or {}).get("function") or {}
        raw_args = fn.get("arguments")
        if isinstance(raw_args, dict):
            args = raw_args
        else:
            try:
                args = json.loads(raw_args or "{}")
            except Exception:
                args = {}
        if not isinstance(args, dict):
            args = {}
        calls.append({
            "id": str((tc or {}).get("id") or "") or f"call_{i}",
            "name": fn.get("name") or "",
            "arguments": args,
        })
    return {
        "choices": [{
            "message": {
                "content": message.get("content") or "",
                "tool_calls": calls,
            }
        }],
        "usage": result.get("usage", {}),
    }

File: Auto_Use/web/llm_provider/together/service.rs
// Copyright 2026 Cursortouch — Auto-Use

//! Together AI provider (OpenAI-compatible chat completions).

use std::sync::atomic::{AtomicBool, Ordering};

use serde_json::{json, Map, Value};

use crate::llm_provider::{api_url, normalize_openai_tool_calls, post_json, SCREENSHOT_MEDIA_TYPE};

pub struct TogetherProvider {
    pub api_key: String,
    pub cli_agent: bool,
    /// OpenAI-format function tools; None only for mode="text".
    pub tools: Option<Vec<Value>>,
    /// Together documents tool_choice as auto|none|{function}; "required" is
    /// undocumented. Start with "required" (the loop's contract) and flip
    /// this for the rest of the session only once Together proves it rejects
    /// "required" — see send_request. Atomic because send_request is &self.
    pub tool_choice_auto: AtomicBool,
}

impl TogetherProvider {
    pub fn send_request(
        &self,
        messages: &[Value],
        model: &str,
        annotated_screenshot_base64: Option<&str>,
    ) -> Result<Value, String> {
        let mut messages: Vec<Value> = messages.to_vec();

        if let Some(shot) = annotated_screenshot_base64 {
            if !self.cli_agent && messages.len() > 1 {
                let last = messages.len() - 1;
                let user_message = messages[last].get("content").cloned().unwrap_or(Value::Null);
                messages[last]["content"] = json!([
                    {"type": "text", "text": user_message},
                    {"type": "image_url",
                     "image_url": {"url": format!("data:{SCREENSHOT_MEDIA_TYPE};base64,{shot}")}},
                ]);
            }
        }

        let mut data = Map::new();
        data.insert("model".into(), json!(model));
        data.insert("messages".into(), Value::Array(messages));
        data.insert("temperature".into(), json!(0.2));
        // Reasoning models bill their thinking as completion tokens; 4000
        // risks finish_reason "length" before the tool call is emitted.
        data.insert("max_tokens".into(), json!(10000));
        let use_auto = self.tool_choice_auto.load(Ordering::Relaxed);
        if let Some(tools) = &self.tools {
            data.insert("tools".into(), Value::Array(tools.clone()));
            // "required" — every turn must call at least one tool; canonical
            // rationale in openrouter/service.rs.
            data.insert("tool_choice".into(), json!(if use_auto { "auto" } else { "required" }));
        }

        // api.together.ai is an alias of the same endpoint.
        let url = api_url("https://api.together.xyz/v1/chat/completions");
        let auth = format!("Bearer {}", self.api_key);
        let headers = [
            ("Authorization", auth.as_str()),
            ("Content-Type", "application/json"),
        ];
        let mut resp = post_json(&url, &headers, &Value::Object(data.clone()))
            .map_err(|e| format!("Together API request failed: {e}"))?;
        if resp.status == 400 && self.tools.is_some() && !use_auto {
            // Retry once with "auto". Remember it ONLY if that works — an
            // unrelated 400 falls through and surfaces its own body. The
            // agent loop repairs the odd text-only turn ("no tool called"),
            // so "auto" degrades gracefully.
            data.insert("tool_choice".into(), json!("auto"));
            if let Ok(retry) = post_json(&url, &headers, &Value::Object(data)) {
                if retry.status < 400 {
                    self.tool_choice_auto.store(true, Ordering::Relaxed);
                    resp = retry;
                }
            }
        }
        if resp.status >= 400 {
            return Err(format!(
                "Together API request failed: HTTP {}\nResponse: {}",
                resp.status, resp.body
            ));
        }
        let result: Value = serde_json::from_str(&resp.body)
            .map_err(|e| format!("Together API request failed: {e}"))?;
        if self.tools.is_some() {
            Ok(normalize_tool_response(&result))
        } else {
            Ok(result)
        }
    }
}

fn normalize_tool_response(result: &Value) -> Value {
    let message = result
        .get("choices")
        .and_then(Value::as_array)
        .and_then(|c| c.first())
        .and_then(|c| c.get("message"))
        .cloned()
        .unwrap_or(json!({}));
    let calls = normalize_openai_tool_calls(&message);
    let content = message.get("content").cloned().unwrap_or(Value::Null);
    json!({
        "choices": [{
            "message": {
                "content": if crate::browser::truthy(&content) { content } else { json!("") },
                "tool_calls": calls,
            }
        }],
        "usage": result.get("usage").cloned().unwrap_or(json!({})),
    })
}
